library/src/image.cxx

```diff
@@ -496,6 +548,42 @@ std::string image::toTerminalText() const
   return ss.str();
 }
 
+//----------------------------------------------------------------------------
+f3d::image& image::setMetadata(const std::string& key, const std::string& value)
+{
+  if (value.empty())
+  {
+    this->Internals->Metadata.erase(key);
+  }
+  else
+  {
+    this->Internals->Metadata[key] = value;
+  }
+  return *this;
+}
+
+//----------------------------------------------------------------------------
+std::string image::getMetadata(const std::string& key) const
+{
+  if (this->Internals->Metadata.count(key))
+  {
+    return this->Internals->Metadata[key];
+  }
+  else
+  {
+    throw std::out_of_range(key);
+  }
+}
+
+//----------------------------------------------------------------------------
+std::set<std::string> image::allMetadata() const
+{
+  std::set<std::string> keys;
```

Comment:
Any reason to use a set instead of a vector here?

Reply:
It makes me feel smart to go "ahah the keys are unique and unordered, they're a _set_ not a _list_" but apart from that not much.
I guess if there were _lots_ of keys a set implementation could be faster but we won't ever have many keys and a vector may even be more performant at the lower count. But it's small potatoes either way.

So up to you, can be a vector if you prefer

Reply:
`std::set` is ordered btw, you may want to use `std::unordered_set` if you really want it to be unsorted.  
However, usually `set` is used when pushing elements in it and we want to make sure it's always sorted and unique. However, insertion complexity is `log(n)`, and while the search is, in theory, faster, `vector` outperforms it with a small number of elements because it is contiguous in memory.  
Sorry to insist, but yeah I'd prefer `std::vector` here.